internal/terraform/rootmodule/root_module.go

```diff
@@ -489,20 +495,42 @@ func (rm *rootModule) parsedFiles() map[string]*hcl.File {
 }
 
 func (rm *rootModule) MergedSchema() (*schema.BodySchema, error) {
+	var err error
+	preloadProviderSchemasOnce.Do(func() {
+		var f http.File
+		f, err = preloadedSchemas.Files.Open("schemas.json")
+		if err != nil {
+			return
+		}
+
+		err = json.NewDecoder(f).Decode(&preloadedProviderSchemas)
+	})
+	if err != nil {
+		rm.logger.Printf("decoding of the preloaded provider schemas has errored and will not be retried: %s", err)
+	}
+
 	var mergedSchema *schema.BodySchema
 
 	if rm.IsCoreSchemaLoaded() {
 		mergedSchema = rm.coreSchema
 	}
 
-	if rm.IsProviderSchemaLoaded() {
-		if !rm.IsParsed() {
-			err := rm.ParseFiles()
-			if err != nil {
-				return nil, err
-			}
+	if !rm.IsParsed() {
+		err := rm.ParseFiles()
+		if err != nil {
+			return nil, err
 		}
-		s, err := tfschema.MergeCoreWithJsonProviderSchemas(rm.parsedFiles(), mergedSchema, rm.providerSchema)
+	}
+
+	ps := &preloadedProviderSchemas
+	if rm.IsProviderSchemaLoaded() {
+		rm.providerSchemaMu.RLock()
+		defer rm.providerSchemaMu.RUnlock()
+		ps = rm.providerSchema
+	}
+
+	if ps == nil {
```

Comment:
I think this will never be true, given that we assign a pointer to `ps` above? 🤔

Reply:
it can be nil if the preloaded schemas fail, we can continue successful operation, but the user has to wait until the actual module schemas are available

Reply:
I would probably expect to see an error if the loading failed, I mean given that we're loading it "from memory" that should _never_ happen, unless something went really wrong, right?

Reply:
It would be odd, I could have it return error, but the problem is since I call it in `Once.Do` the function will only error the first time. That's why I chose to actually just log the error. I could panic on failure? It wouldn't be a bad case for it, as the program really shoudn't continue

Reply:
Hmm, I see - how about persisting that error in a global variable such as `preloadedProviderSchemasErr` and _always_ returning it?

Reply:
👍